src/client.cpp

```diff
@@ -291,31 +281,89 @@ void CClient::OnNewConnection()
     //### TODO: END ###//
 }
 
-void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+void CClient::OnMuteStateHasChangedReceived ( int iServerChanID, bool bIsMuted )
+{
+    // map iChanID from server channel ID to client channel ID
+    int iChanID = FindClientChannel ( iServerChanID, false );
+
+    emit MuteStateHasChangedReceived ( iChanID, bIsMuted );
+}
+
+void CClient::OnCLChannelLevelListReceived ( CHostAddress InetAddr, CVector<uint16_t> vecLevelList )
 {
-    // Upon receiving a new client list, we have to reset oldGain and newGain
-    // entries for unused channels. This ensures that a disconnected channel
-    // does not leave behind wrong cached gain values which would leak into
-    // any new channel which reused this channel id.
-    int iNumConnectedClients = vecChanInfo.Size();
+    // reorder levels from server channel order to client channel order
 
-    // Save what channel IDs are in use:
-    bool bChanIdInUse[MAX_NUM_CHANNELS] = {};
-    for ( int i = 0; i < iNumConnectedClients; i++ )
+    if ( ReorderLevelList ( vecLevelList ) )
     {
-        bChanIdInUse[vecChanInfo[i].iChanID] = true;
+        emit CLChannelLevelListReceived ( InetAddr, vecLevelList );
     }
+}
+
+void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+{
+    // translate from server channel IDs to client channel IDs
+    // ALSO here is where we allocate and free client channels as required
 
-    // Reset all gains for unused channel IDs:
-    for ( int iId = 0; iId < MAX_NUM_CHANNELS; iId++ )
+    const int iNumConnectedClients = vecChanInfo.Size();
+    int       i, iSrvIdx;
+
+    // on a new connection, a server sends an empty channel list before sending
+    // the real channel list (see #1010). To avoid this discarding "our" channel
+    // that we have just created, we skip this processing and just pass the empty
+    // list to the emitted signal.
+
+    if ( iNumConnectedClients != 0 )
     {
-        if ( !bChanIdInUse[iId] )
+        // this relies on the received client list being in order of server channel ID
+
+        for ( i = 0, iSrvIdx = 0; i < iNumConnectedClients && iSrvIdx < MAX_NUM_CHANNELS; )
         {
-            // reset oldGain and newGain as this channel id is currently unused and will
-            // start with a server-side gain at 1 (100%) again.
-            oldGain[iId] = newGain[iId] = 1;
+            // server channel ID of this entry
+            const int iServerChannelID = vecChanInfo[i].iChanID;
+
+            // find matching client channel ID, creating new if necessary
+            const int iClientChannelID = FindClientChannel ( iServerChannelID, true );
```

Comment:
I think I'd be happier if this were `iChanID` -- we're going to store it in `vecChanInfo[i].iChanID` later.

Talking about `vecChanInfo[i].iChanID` as `iServerChannelID` and `iClientChannelID` is a bit confusing.

Reply:
In fact, between here and line 341, I don't see `i` or `iClientChannelID` or `vecChanInfo` changing -- why can't it be stored now?  (Presumably `FindClientChannel` is hiding something that should be visible here somehow.)

Reply:
Yes, in fact we don't need `iClientChannelID` at all. Line 325 could just become:

```
vecChanInfo[i].iChanID = FindClientChannel ( iServerChannelID, true );
```

With a suitably-expanded comment...